src/conn/conn_api.c

```diff
@@ -2621,6 +2627,54 @@ __conn_chk_file_system(WT_SESSION_IMPL *session, bool readonly)
     return (0);
 }
 
+/*
+ * __conn_config_file_system --
+ *     Configure the file system on the connection if the user hasn't added a custom file system.
+ */
+static int
+__conn_config_file_system(WT_SESSION_IMPL *session, const char *cfg[])
+{
+    WT_CONFIG_ITEM cval;
+    /*
+     * If the application didn't configure its own file system, configure one of ours. Check to
+     * ensure we have a valid file system.
+     *
+     * Check the "live_restore" config. If it is provided validate that a custom file system has not
+     * been provided, and that the connection is not in memory or Windows.
+     */
+    WT_RET(__wt_config_gets(session, cfg, "live_restore.enabled", &cval));
+
+    WT_CONNECTION_IMPL *conn = S2C(session);
+    if (cval.val) {
+        F_SET(conn, WT_CONN_LIVE_RESTORE);
+        /* Live restore compatibility checks. */
+        if (conn->file_system != NULL)
+            WT_RET_MSG(session, EINVAL, "Live restore is not compatible with custom file systems");
+        if (F_ISSET(conn, WT_CONN_IN_MEMORY))
+            WT_RET_MSG(
+              session, EINVAL, "Live restore is not compatible with an in-memory connections");
+#ifdef _MSC_VER
+        WT_RET_MSG(session, EINVAL, "Live restore is not supported on Windows");
+#endif
+    }
+
+    if (conn->file_system == NULL) {
```

Comment:
Given also the check above, when enabled, of `conn->file_system != NULL`, can we move that check to the top and do an early return and get rid of a few conditionals and indentation levels?

Reply:
This speaks to a bigger problem within WiredTiger where we don't have a great way of dealing with incompatible configuration. The first check against it validates that it doesn't clash with live restore, the second check is a catch all that ensures we don't create a file system if there already is one. I'm guessing here but I think the second file_system check was added a long time ago and whoever put it in didn't add full compatibility checking. There should be a message like "if the is a file system the connection cannot be in memory too".

Without fixing the bigger problem I don't think I can tease apart these if statements.

Reply:
Fair enough. Thanks for thinking about it.